pop-api/examples/fungibles/tests.rs

```diff
@@ -0,0 +1,831 @@
+		Err(into_psp22_custom(Module { index: 52, error: [16, 0] }))
+	);
+}
+
+/// 4. PSP-22 Burnable Interface:
+/// - burn
+
+#[drink::test(sandbox = Sandbox)]
+fn burn_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// No permission to burn.
+	assert_ok!(session.sandbox().create(&(TOKEN + 1), &BOB, MIN_BALANCE)); // Create a new token owned by `BOB`.
+	assert_ok!(session.sandbox().mint_into(&(TOKEN + 1), &BOB, AMOUNT));
+	assert_ok!(deploy(&mut session, CONTRACT.clone(), "existing", vec![(TOKEN + 1).to_string()]));
+	// `pallet-assets` returns `NoPermission` error.
+	assert_eq!(
+		burn(&mut session, BOB, AMOUNT),
+		Err(into_psp22_custom(Module { index: 52, error: [2, 0] }))
+	);
+
+	// Deploy a new contract.
+	let contract = deploy(
+		&mut session,
+		CONTRACT.clone(),
+		"new",
+		vec![TOKEN.to_string(), MIN_BALANCE.to_string()],
+	)
+	.unwrap();
+	let value = 1;
+	session.set_actor(contract.clone());
+	// Mint tokens.
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &ALICE, AMOUNT));
+
+	// No-op.
+	assert_ok!(burn(&mut session, ALICE, 0));
+	assert_eq!(last_contract_event(&session), None); // No event emitted.
+
+	// Failed with `InsufficientBalance`.
+	assert_eq!(burn(&mut session, ALICE, AMOUNT * 2), Err(PSP22Error::InsufficientBalance));
+
+	// Successfully burn tokens.
+	assert_ok!(burn(&mut session, ALICE, value));
+	assert_eq!(session.sandbox().total_supply(&TOKEN), AMOUNT - value);
+	assert_eq!(session.sandbox().balance_of(&TOKEN, &ALICE), AMOUNT - value);
+	// Successfully emit event.
+	assert_eq!(
+		last_contract_event(&session).unwrap(),
+		Transfer { from: Some(account_id_from_slice(ALICE.as_ref())), to: None, value }
+			.encode()
+			.as_slice()
+	);
+
+	// Token is not live, i.e. frozen or being destroyed.
+	assert_ok!(session.sandbox().start_destroy(&TOKEN));
+	// `pallet-assets` returns `IncorrectStatus` error.
+	assert_eq!(
+		burn(&mut session, ALICE, value),
+		Err(into_psp22_custom(Module { index: 52, error: [17, 0] }))
+	);
+}
+
+/// A set of helper methods to test the contract calls.
+mod utils {
+	use drink::session::{error::SessionError, NO_SALT};
+	use pop_api::primitives::AccountId;
+	use pop_sandbox::{AccountId32, Balance, INIT_VALUE};
+	use scale::Decode;
+
+	use super::*;
+
+	/// Deploy test utilities.
+
+	pub(super) fn deploy(
+		session: &mut Session<Sandbox>,
+		bundle: ContractBundle,
+		method: &str,
+		inputs: Vec<String>,
+	) -> Result<AccountId32, PSP22Error> {
+		let result = session.deploy_bundle(bundle, method, &inputs, NO_SALT, Some(INIT_VALUE));
+		if result.is_err() {
+			let deployment_result = session.record().last_deploy_result().result.clone();
+			let error = deployment_result.unwrap().result.data;
+			return Err(PSP22Error::decode(&mut &error[2..]).unwrap());
+		}
+		Ok(result.unwrap())
+	}
+
+	/// PSP22 test utilities.
+
+	pub(super) fn total_supply(session: &mut Session<Sandbox>) -> Balance {
+		call::<Balance>(session, "PSP22::total_supply", vec![], None).unwrap()
```

Comment:
Is there a way to get this from the contract in a type safe way? That is, does the message macro perhaps expose a function which determines this name.

Some of the identation should also be removed, hopefully reducing the lines of code and making it easier to absorb for devs.

Reply:
Not sure how we can do this in the test. The `<TraitName>::<FunctionName>` is derived from the AST handled inside the macro for a contract message.